src/Orleans.Runtime/Scheduler/WorkItemGroup.cs

```diff
@@ -211,7 +232,7 @@ public void EnqueueTask(Task task)
 #if DEBUG
                 if (log.IsEnabled(LogLevel.Trace)) log.Trace("Add to RunQueue {0}, #{1}, onto {2}", task, thisSequenceNumber, SchedulingContext);
 #endif
-                masterScheduler.RunQueue.Add(this);
+                masterScheduler.ScheduleExecution(this);
```

Comment:
Do we still need to enqueue this if it's already enqueued, i.e, `state == Runnable`?
It seems like we can just rely on the fact that it will be re-enqueued at the end of its `Execute()` method if there are work items in the local queue.

Reply:
The execution of `WorkItemGroup`  is not scheduled at this moment. That is because of non atomic switch to `Runnable` state. That state transition might be more explicit if extracted into separate method (e.g. as on  https://github.com/dotnet/orleans/pull/2060/files#diff-21a21498604dbe8a2a72f61a4f1a628cR424)

Reply:
As for consideration of single queue being a point of contention for all activations - that can be addressed with thread local queues (which already exists in .NET ThreadPool).

Reply:
>The execution of WorkItemGroup is not scheduled at this moment.

oh, I missed the check above